Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,99 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+fileprivate enum ClientState: Hashable {
+    case ready
+    case waitingForAuthenticationMethod(ClientGreeting)
+    case waitingForClientRequest
+    case waitingForServerResponse(ClientRequest)
+    case active
+}
+
+enum ClientAction: Hashable {
+    case none
+    case sendRequest
+    case proxyEstablished
+}
+
+public struct InvalidAuthenticationSelection: Error {
+    
+}
+
+public struct ConnectionFailed: Error, Hashable {
+    public var reply: Reply
+}
+
+struct ClientStateMachine {
+
+    private var state: ClientState
```

Comment:
Do we maybe want to move the `ClientState` into here?
```suggestion
    private enum State: Hashable {
        case ready
        case waitingForAuthenticationMethod(ClientGreeting)
        case waitingForClientRequest
        case waitingForServerResponse(ClientRequest)
        case active
    }

    private var state: State
```

Reply:
I think not personally - do you feel strongly on this?

Reply:
Not really! Resolve... Purely personal opinion.